File: control-plane/api-gateway/src/routes/events.rs
use axum::{
    body::Body,
    extract::{Path, RawQuery, State},
    http::{HeaderMap, StatusCode},
    response::{IntoResponse, Json},
    routing::{delete, get, post},
    Router,
};
use serde_json::json;
use uuid::Uuid;

use crate::{auth::middleware::AuthenticatedUser, AppState};

async fn proxy_to_failover(
    state: &AppState,
    method: reqwest::Method,
    path: &str,
    body: Option<serde_json::Value>,
    headers: Option<Vec<(String, String)>>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    match state
        .service_client
        .forward_to_failover_controller(method, path, body, headers)
        .await
    {
        Ok((status, Some(body))) => {
            let axum_status =
                StatusCode::from_u16(status.as_u16()).unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
            Ok((axum_status, Json(body)).into_response())
        }
        Ok((status, None)) => {
            let axum_status =
                StatusCode::from_u16(status.as_u16()).unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
            Ok((axum_status, Body::empty()).into_response())
        }
        Err(e) => {
            tracing::error!("Failed to forward request to failover-controller: {}", e);
            Err((
                StatusCode::BAD_GATEWAY,
                Json(
                    json!({ "error": "Failover controller unavailable", "details": e.to_string() }),
                ),
            ))
        }
    }
}

pub async fn list_events(
    AuthenticatedUser(_claims): AuthenticatedUser,
    State(state): State<AppState>,
    headers: HeaderMap,
    RawQuery(query): RawQuery,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let header_map = headers
        .iter()
        .filter_map(|(k, v)| v.to_str().ok().map(|val| (k.to_string(), val.to_string())))
        .collect();
    let path = match query {
        Some(q) => format!("/events?{}", q),
        None => "/events".to_string(),
    };
    proxy_to_failover(&state, reqwest::Method::GET, &path, None, Some(header_map)).await
}

pub async fn set_maintenance(
    AuthenticatedUser(_claims): AuthenticatedUser,
    State(state): State<AppState>,
    Path(agent_id): Path<Uuid>,
    headers: HeaderMap,
    Json(body): Json<serde_json::Value>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let header_map = headers
        .iter()
        .filter_map(|(k, v)| v.to_str().ok().map(|val| (k.to_string(), val.to_string())))
        .collect();
    proxy_to_failover(
        &state,
        reqwest::Method::POST,
        &format!("/agents/{}/maintenance", agent_id),
        Some(body),
        Some(header_map),
    )
    .await
}

pub async fn clear_maintenance(
    AuthenticatedUser(_claims): AuthenticatedUser,
    State(state): State<AppState>,
    Path(agent_id): Path<Uuid>,
    headers: HeaderMap,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let header_map = headers
        .iter()
        .filter_map(|(k, v)| v.to_str().ok().map(|val| (k.to_string(), val.to_string())))
        .collect();
    proxy_to_failover(
        &state,
        reqwest::Method::DELETE,
        &format!("/agents/{}/maintenance", agent_id),
        None,
        Some(header_map),
    )
    .await
}

pub fn events_routes() -> Router<AppState> {
    Router::new()
        .route("/events", get(list_events))
        .route("/agents/{agent_id}/maintenance", post(set_maintenance))
        .route("/agents/{agent_id}/maintenance", delete(clear_maintenance))
}

File: app/src/lib/api/events.ts
import { authedFetch } from './http';

const API_BASE = '/api';

export interface AlertEvent {
    id: string;
    agent_id: string | null;
    event_type: string;
    affected_workloads: string[] | null;
    duration_ms: number | null;
    created_at: string;
    severity: string;
    status: string;
    resolved_at: string | null;
    message: string | null;
}

export async function listEvents(
    token: string,
    filters?: { status?: string; agentId?: string },
): Promise<AlertEvent[]> {
    const params = new URLSearchParams();
    if (filters?.status) params.set('status', filters.status);
    if (filters?.agentId) params.set('agent_id', filters.agentId);
    const query = params.toString();
    const res = await authedFetch(`${API_BASE}/events${query ? `?${query}` : ''}`, {
        headers: { Authorization: `Bearer ${token}` },
    });
    if (!res.ok) throw new Error(`events fetch failed: ${res.status}`);
    return res.json();
}

export async function setMaintenance(token: string, agentId: string, minutes: number): Promise<void> {
    const res = await authedFetch(`${API_BASE}/agents/${agentId}/maintenance`, {
        method: 'POST',
        headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
        body: JSON.stringify({ minutes }),
    });
    if (!res.ok) {
        const err = await res.json().catch(() => ({ error: res.status }));
        throw new Error(err.error ?? `Failed to set maintenance: ${res.status}`);
    }
}

export async function clearMaintenance(token: string, agentId: string): Promise<void> {
    const res = await authedFetch(`${API_BASE}/agents/${agentId}/maintenance`, {
        method: 'DELETE',
        headers: { Authorization: `Bearer ${token}` },
    });
    if (!res.ok) {
        const err = await res.json().catch(() => ({ error: res.status }));
        throw new Error(err.error ?? `Failed to clear maintenance: ${res.status}`);
    }
}
